pkg/nsx/services/subnet/subnet.go

```diff
@@ -428,36 +438,39 @@ func (service *SubnetService) UpdateSubnetSet(ns string, vpcSubnets []*model.Vpc
 		}
 		newTags := append(service.buildBasicTags(subnetSet), tags...)
 
-		var updatedSubnet *model.VpcSubnet
+		var comparableSubnet *model.VpcSubnet
 		if vpcSubnets[i].SubnetDhcpConfig == nil {
-			updatedSubnet = &model.VpcSubnet{
+			comparableSubnet = &model.VpcSubnet{
 				Tags: newTags,
 			}
 		} else {
-			updatedSubnet = &model.VpcSubnet{
+			comparableSubnet = &model.VpcSubnet{
 				Tags:             newTags,
 				SubnetDhcpConfig: service.buildSubnetDHCPConfig(dhcpMode),
 			}
 		}
-		changed := common.CompareResource(SubnetToComparable(vpcSubnets[i]), SubnetToComparable(updatedSubnet))
+		changed := common.CompareResource(SubnetToComparable(vpcSubnets[i]), SubnetToComparable(comparableSubnet))
 		if !changed {
 			log.Info("NSX Subnet unchanged, skipping update", "Subnet", *vpcSubnet.Id)
 			continue
 		}
 
-		vpcSubnets[i].Tags = newTags
+		// Avoid updating vpcSubnets[i] to ensure Subnet store
+		// is only updated after the updating succeeds.
+		updatedSubnet := *vpcSubnets[i]
```

Comment:
Why we need two new variables `comparableSubnet` and `updatedSubnet` here

Reply:
It makes sense to merge them, use only updatedSubnet instead, thanks!